### PR TITLE
Issue #4216: close LiCCA full-suite gate

### LINKED ISSUE #4216: tests: 15 LiCCA-environment-conditional failures in the full suite (env guards/skips needed)
## Context
LiCCA full-suite run 9858988 (2026-07-03, full extras incl. orca after the #4192/#4197 self-heal fix): **10310 passed, 15 failed, 11 skipped, 1 error** in 13 min — vs the prior no-orca run's 9900/93/379. The suite is effectively green on LiCCA; the residue is environment-conditional.

## The 15 failures (from pytest.log, job 9858988)
- `tests/dev/test_pr_ready_preflight.py` (base-ref/module assumptions that don't hold on a cluster worktree)
- `tests/dev/test_ci_uv_sync_diag.py` (CI-runner introspection expects GH-runner env)
- `tests/test_visualization_performance.py` (wall-clock threshold on a shared 128-core node)
- (full list in `/hpc/gpfs2/scratch/u/luttkule/robot_sf/licca-all-tests/9858988/pytest.log`)

## Acceptance criteria
- Each failing test gets an explicit environment guard (skip with reason) OR an env-agnostic rewrite — decided per test, not blanket-skipped.
- The guards must NOT weaken CI on GitHub runners (assert the tests still run there).
- Rationale comment per guard naming the env assumption.

Payoff: LiCCA full-suite becomes a clean 0-failure post-merge validation gate for cluster campaigns.

### PR DESCRIPTION
**Pull request by ll7**
## Reviewer-critical summary

What changed: closes the remaining #4216 validation gate by adding a LiCCA/simulated-LiCCA pytest session-finish cleanup for leaked nested pytest descendants, making behavior-token quantization deterministic under `ROBOT_SF_TEST_ENV=licca`, and recording the final green full-suite evidence in the canonical #4216 state surface.

Why now: the live issue thread and merged PRs #4237/#4276/#4320/#4340/#4343/#4513/#4575 showed 7 of 8 criteria complete; the only remaining close blocker was a full-suite `ROBOT_SF_TEST_ENV=licca` exit-code rerun with a final pytest summary.

Claim boundary: #4216 closure gate only. The decisive full-suite proof ran on `imech036` local validation with simulated LiCCA env, not Slurm/GPU compute-submit. No benchmark campaign, release, merge, deletion, or paper/dissertation claim edit.

Required validation: focused teardown/quantizer tests, docs/evidence consistency, and full simulated-LiCCA suite. Full-suite proof: `11941 passed, 16 skipped, 11 warnings in 1243.30s (0:20:43)` on validation base `ac309f874`; after final rebase to `972ea8235`, focused Ruff/tests/docs checks passed.

Remaining blocker: none for #4216. `docs/context/issue_4216_state.yaml` now records `status: green_gate` and `remaining_empirical_action.blocks_close: false`.

## Contract delta

New schema fields: none for runtime APIs. Existing issue-state/evidence documents add a final green-gate entry and compact evidence JSON.

New fail-closed blockers: none. The prior hang-regression blocker is converted to a green gate by the final full-suite summary.

Compatibility impact: test-infrastructure cleanup is gated to `ROBOT_SF_TEST_ENV=licca` or explicit `ROBOT_SF_REAP_LEAKED_PYTEST_CHILDREN=1`; normal local/GitHub sessions are not broadened. Behavior-token quantization now uses the existing deterministic NumPy implementation instead of scikit-learn KMeans so seeded assignments are stable on degenerate inputs.

## Issue

Closes #4216.

## Scope summary

- Added `reap_matching_descendants()` in `tests/support/process_teardown.py` and wired it from `tests/conftest.py` for LiCCA/simulated-LiCCA teardown cleanup.
- Switched behavior-token quantization to deterministic NumPy k-means and stable canonical-center assignment.
- Added `docs/context/evidence/issue_4216_licca_full_suite_env_failures/final_green_gate_20260706.json` and appended a final acceptance-evidence table to `docs/context/issue_4216_state.yaml`.

## Validation command results

| Command | Exit | Result |
| --- | ---: | --- |
| `PYTEST_NUM_WORKERS=8 ROBOT_SF_TEST_ENV=licca python scripts/dev/run_compact_validation.py --artifact-dir .../full-suite-licca-rebased-final-20260706 --timeout-seconds 3600 --json -- scripts/dev/run_worktree_shared_venv.sh -- uv run pytest -q` | 0 | `11941 passed, 16 skipped, 11 warnings in 1243.30s (0:20:43)`; no timeout, no failing node ids |
| `ROBOT_SF_TEST_ENV=licca scripts/dev/run_worktree_shared_venv.sh -- uv run pytest tests/support/test_process_teardown.py tests/experiments/test_behavior_tokens.py -q` | 0 | `12 passed` after final rebase |
| `scripts/dev/run_worktree_shared_venv.sh -- uv run ruff check experiments/behavior_tokens/quantize_trace_windows.py tests/experiments/test_behavior_tokens.py tests/support/process_teardown.py tests/support/test_process_teardown.py tests/conftest.py` | 0 | all checks passed after final rebase |
| `scripts/dev/run_worktree_shared_venv.sh -- uv run python scripts/validation/check_docs_proof_consistency.py --check-evidence-catalog` | 0 | evidence catalog coverage passed |
| `python - <<'PY' ... yaml/json parse assertions ... PY` | 0 | state/evidence parse ok |
| `git diff --check` | 0 | no whitespace errors |

## Out-of-scope

No full benchmark campaign run, no Slurm/GPU submission, no paper/dissertation claim edits. No issue/PR comment was posted; propagation is through the single canonical state file because this is a high-churn closure issue and the lane was not authorized for comments.

<details>
<summary>Acceptance evidence</summary>

| Criterion | Evidence | Status |
| --- | --- | --- |
| Full issue thread including comments and linked/merged PRs reviewed | Live issue body and comments read via GitHub API on 2026-07-06; merged #4237/#4276/#4320/#4340/#4343/#4513/#4575 reviewed | Met |
| All 15 job-9858988 failures classified | #4237 merged `failure_inventory.json` | Met |
| GitHub Actions coverage preserved | #4237 merged guard helper/tests proving GitHub Actions wins over LiCCA markers | Met |
| Each failing node has env-agnostic rewrite or narrow guard | #4237 updated affected dev/CI/visualization tests and inventory mappings | Met |
| Post-guard full-suite residue probed fail-closed | #4276 recorded 100%-then-hang as non-closure evidence | Met |
| Nested-pytest teardown hang bounded/reaped | #4320 bounded the known call site; this PR adds LiCCA session-finish descendant cleanup; final suite emitted a normal summary | Met |
| Canonical issue state records closure chain | #4343 created state; #4513/#4575 appended prior audits; this PR appends final green gate | Met |
| New LiCCA/simulated-LiCCA full-suite exit-code rerun confirms closure | Final run exited 0 with `11941 passed, 16 skipped, 11 warnings` | Met |

</details>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved test-run stability in the LiCCA environment by cleaning up leftover child processes after test sessions.
  * Made token quantization behavior deterministic, helping keep results consistent across runs.
  * Updated recorded validation status to reflect a successful full-suite run with no remaining failures.

### CODE REVIEW

**Comment by ll7** [comment]
## Gate review — PR #4694 (REVIEW → IMPROVE/FIX → MERGE)

Reviewed the actual diff (test-teardown infra + deterministic behavior-token quantizer + #4216 closure evidence). Enumerating **every open review thread** with per-thread verdict, plus my own findings.

### Open review threads (CodeRabbit)

1. **[Major, data integrity] Final pytest summary conflicts with `issue_4216_state.yaml`.** — **FIXED.** The decisive green-gate proof is `11941 passed, 16 skipped, 11 warnings in 1243.30s` in the tracked evidence JSON, PR body, the state entry's own `final_summary`, and `recorded_result`. One acceptance-criterion evidence line carried a stale `11912 … 1224.36s`. Aligned it to `11941 … 1243.30s` so the closure evidence is internally consistent.
2. **[Major, data integrity] Acceptance-evidence figures don't match the entry's own evidence block (lines 288-293).** — **FIXED** by the same edit; this was the same `11912/1224.36s` outlier.
3. **[Minor, correctness] Fragile `/proc/<pid>/stat` parsing + unmanaged file handles.** — **FIXED.** `split()[2]` mis-parses when `comm` contains spaces/parens; now parses the state char after the final `)` via `rsplit(")", 1)`. Also wrapped the procfs reads in `_pid_alive`, `_direct_child_pids`, and `_cmdline` (and the test-file copy) in context managers. Added `IndexError` to the guarded exceptions.
4. **[Major, stability] Broaden exception handling in `_terminate_pids`.** — **FIXED.** Both `os.kill` calls now also swallow non-`ProcessLookupError` `OSError` (e.g. `PermissionError`) so cleanup can't propagate an exception out of the unguarded `pytest_sessionfinish` hook.
5. **[Minor, correctness] Race reading `grandchild.pid` before it's fully written.** — **FIXED.** The parent now polls until the pid file holds a fully-written integer (`isdigit()`) instead of trusting `exists()`, eliminating the empty-read `ValueError` flake.

**Open review threads remaining: 0 unaddressed** (5 total, all fixed in phase 2).

### Gate's own findings

6. **Behavior-token quantizer switches from sklearn KMeans to always-NumPy k-means.** Reviewed as a deliberate determinism change confined to `experiments/behavior_tokens/` (not a core benchmark metric path); within the PR's stated claim boundary and covered by `test_quantizer_is_deterministic`. The new `_assign_to_centers` reassigns to canonical centers by nearest distance, which is equivalent to the converged k-means assignment plus deterministic tie-breaking. No action.

### Validation (post-fix, this worktree)
- `ruff check` on all changed files → passed
- `ROBOT_SF_TEST_ENV=licca pytest tests/support/test_process_teardown.py tests/experiments/test_behavior_tokens.py` → `12 passed`
- state.yaml/evidence JSON parse ok; `git diff --check` clean

The decisive full-suite green gate (`11941 passed … 1243.30s`, exit 0) is unchanged; my edits are teardown-robustness + evidence-consistency only and do not alter that result.
